### DISCUSSION TITLE
Calculating 95th percentiles in PointStat

### DISCUSSION THREAD

**Discussion by DeannaSpindler-NOAA** [dtcenter/METplus#1966] [Statistical Computation] [answered]
Hello,
Is it possible to get the 95th percentile in PointStat? I know I can get the 90th percentile using the CNT line, but how can I specify the 95th percentile?
Thanks,
Deanna

**Comment by JohnHalleyGotway** [comment]
Hi @DeannaSpindler-NOAA, I see you have a question about the percentiles of the fcst - obs differences reported in the [CNT line type](https://met.readthedocs.io/en/latest/Users_Guide/point-stat.html?highlight=CNT#id11). Columns 81 to 94 contain the 10th, 25th, 50th, 75th, and 90th percentiles of those differences.

However you're interested in the 95th percentile, and I could imagine that other users might be interested in the 99th, 1st, or 5th percentiles as well. Unfortunately, there is no clean and simple direct solution in the current version.

Should we be reporting more percentiles of the error here? Or should we make the percentiles to be reported configurable?

This reminds me of similar functionality elsewhere:

1. The `stat_analysis` summary job type also reports percentiles, and can be run on the MPR line type produced by Point-Stat to summarize the `FCST-OBS` differences:
```
stat_analysis -lookin out/point_stat/point_stat_360000L_20070331_120000V_mpr.txt -job summary -line_type MPR -column FCST-OBS
```
However, the output of the [summary job](https://met.readthedocs.io/en/latest/Users_Guide/stat-analysis.html#summary-statistics-for-columns) includes the same percentiles as the CNT line type.

2. In the [MODE configuration file](https://met.readthedocs.io/en/latest/Users_Guide/mode.html#mode-configuration-file), the `inten_perc_value` specifies the intensity percentile within the object that the user cares about. And that's used in the `inten_perc_ratio` weight and reported to the output. But that's the intensity of values within the object rather than the FCST - OBS differences.

3. Likewise, you can define thresholds relative to the distribution of forecast, observation, or climatological values. However those [percentile thresholds](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=SFP#configuration-file-overview) are computed over a single dataset, not the FCST - OBS differences.

For now, the only way to do this would be outside of MET. Configure Point-Stat to write the MPR line type, and then use python (or something else) to read those MPR lines, compute the FCST-OBS difference, and report the 95th percentile. Or you could change one line of code in a custom version to report the 95th percentile instead of the 90th percentile... but that isn't a good longterm solution. FYI, here's the [line of code](https://github.com/dtcenter/MET/blob/2bc783e4c6690a45f927b64b66e6f1ad927f8ba8/src/libcode/vx_statistics/compute_stats.cc#L307) where the 90th percentile is computed. Changing `0.90` to `0.95` would produce the desired outcome.

I wonder if/how we should support this directly in MET in the future. It's obviously trivial to compute, but deciding the mechanics is the hard part.

Should we could add more column(s) to the CNT line type for pre-defined percentiles of the error? Should we make the 5 existing columns configurable? If so, how do we indicate the actual percentile value in the output? Should we add an entirely new line type with more info about the distribution of differences?

**Comment by DeannaSpindler-NOAA** [comment]
Hello John,

Thank you for answering my question.

I'll defer to you/DTC as to how to proceed, but I know for wave models it
is very interesting to track the highest percentiles (not just the 95th)
because while the 95th percentile of global significant wave heights is
around 3 m, most everyone is interested in the wave heights higher than 4m,
8m, etc.  In my case here, I was looking to compare the 95th percentile of
the forecasts compared to that of the observations (and maybe the
climatology) at the matched locations - to see if the model is capturing
the highest wave heights, over or under-estimating them, and so on.  I'm
interested in the differences between the model and observations at the
highest significant wave heights - not really in the distribution of the
differences.  That might be a new line type I think (up to DTC's
discretion), for which I would vote for at least two configurable
percentile columns.

Thanks,
Deanna

--
Deanna Spindler, PhD
Physical Scientist III
Lynker at NOAA/NWS/NCEP/Environmental Modeling Center
NOAA Center for Weather and Climate Prediction

**Reply by JohnHalleyGotway** [reply]
@DeannaSpindler-NOAA, OK, thanks for clarifying. Based on this, I definetely think you should make use of the [percentile threshold](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=SFP#configuration-file-overview) functionality in MET. They are an excellent fit for this type of question.

And you can use them for categorical verification or for continuous verification.

For example, you could evaluate wave height using `cat_thresh = [ >SFP90, >SFP95, >SFP99, >SOP90, >SOP95, >SOP99 ];` and then write out the contingency table counts (CTC) and statistics (CTS). After creating the matched pairs, these settings will extract the 90/95/99-th percentiles from the distribution of forecast and observation values. Then it'll define thresholds based on those and compute/write contingency table output for each threshold listed.

It's up to you to define whether you care about percentiles of the forecast values or observation values. Both are potentially useful but answer different scientific questions.

You can also use the same thresholds to set `cnt_thresh =  [ >SFP90, >SFP95, >SFP99, >SOP90, >SOP95, >SOP99 ];` and then write partial sums (SL1L2) and continuous statistics (CNT). The `cnt_thresh` option *FILTERS* the matched pairs and only keeps the ones that meet the threshold criteria. So you'll get RMSE (for example) computed over subsets of the matched pairs where the threshold criteria is met. I'd expect that you'd see larger errors for more extreme events.

Lastly, sometime folks are uncomfortable using percentile thresholds entirely since they are defined only on the current sample of data. I really like the `==FBIAS1.0` threshold option since it gives you a really nice combination of real world thresholds and percentile thresholds.

For example, in the `obs` dictionary, you could set `cat_thresh = [ ge4, get5, ge8 ];` and in the `fcst` dictionary, set `cat_thresh = [ ==FBIAS1, ==FBIAS1, ==FBIAS1 ];`. For each observation threshold, MET figures out how many of the obs values meet that criteria. Then it automatically selects a forecast threshold value that will yield the same number of points. You're basically telling MET to pick a threshold that will yield a frequency bias value as close to 1 as possible. So it's de-biasing the forecast on the fly.

Please give these options some thought and let me know if that's really what you're looking for.

**Comment by DeannaSpindler-NOAA** [comment]
Hello John,

Can I specify a combination of values and percentiles in the threshold
function? For example:
BOTH_VAR1_NAME = HTSGW
BOTH_VAR1_LEVELS = L0
BOTH_VAR1_THRESH = le3.0, ge4.0, ge8.0, geP95

Thanks,
Deanna

--
Deanna Spindler, PhD
Physical Scientist III
Lynker at NOAA/NWS/NCEP/Environmental Modeling Center
NOAA Center for Weather and Climate Prediction

**Reply by JohnHalleyGotway** [reply]
Deanna,

Yes, you certainly can mix/match threshold types. However, be aware the `geP95` is NOT a valid percentile threshold type. Instead, we need more information to specify the data over which the percentiles should be computed.

Search for "percentile threshold" in [this section](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=percentile%20thresholds#configuration-file-overview) of the user's guide.
 - `geSFP95` uses the 95th percentile of the current sample of forecast values.
 - `geSOP95` uses the 95th percentile of the current sample of observation values.
 - `geSCP95` uses the 95th percentile of the current sample of climatological mean values.

If you want to compare the 95th percentile forecast threshold to the 95th percentile observation threshold, you'll need to specify them separately in the METplus config file:
```
FCST_VAR1_THRESH = le3.0, ge4.0, ge8.0, geSFP95
OBS_VAR1_THRESH = le3.0, ge4.0, ge8.0, geSOP95
```

It's really up to you how you want to select the data over which you want to compute the percentile, and then choose to apply that to threshold the forecast data and/or observation data.

**Comment by DeannaSpindler-NOAA** [comment]
Thanks, John! I'll try specifying them separately.

Best,
Deanna

**Comment by DeannaSpindler-NOAA** [comment]
Is it possible to get the 95th percentile of the current sample of
climatological mean values (geSCP95) output too? How do I ask for that?

**Reply by JohnHalleyGotway** [reply]
Yes, definitely, you can list as many threshold types as you'd like:
```
FCST_VAR1_THRESH = le3.0, ge4.0, ge8.0, geSFP95, geSCP90, geSCP95, geCDP90, geCDP95, ge==FBIAS1, ==FBIAS1, ==FBIAS1
OBS_VAR1_THRESH = le3.0, ge4.0, ge8.0, geSOP95, geSCP90, geSCP95, geCDP90, geCDP95, le3.0, ge4.0, ge8.0
```
You just need to make sure the lengths of the lists match. Above, I've added...
- Thresholding both using the 90th and 95th percentiles of the climo mean (SCP).
- Thresholding both using the 90th and 95th climatological distribution percentile (CDP). This requires that both the climo mean and standard deviation be defined. It lets the threshold vary for each individual grid point based on the climo mean and stdev at that point. Assuming normality, it select the threshold value so that the specified percentage (90 or 95) of the area under the curve is to the left of that point.
- Thresholding the observations using physical values (3.0, 4.0, 8.0) and then letting the forecast threshold vary to produce a frequency bias of 1.0 (==FBIAS1).

Just read about these options in [this section](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=percentile%20thresholds#configuration-file-overview) of the user's guide.

**Comment by DeannaSpindler-NOAA** [comment]
Hello John,

I tried to ask for the different percentiles:

FCST_VAR1_NAME = HTSGW
FCST_VAR1_LEVELS = L0
FCST_VAR1_THRESH = le3.0, ge4.0, ge8.0, geSFP95, geSCP95, geCDP95,
ge==FBIAS1, ==FBIAS1, ==FBIAS1

OBS_VAR1_NAME = HTSGW
OBS_VAR1_LEVELS = L0
OBS_VAR1_THRESH = le3.0, ge4.0, ge8.0, geSOP95, geSCP95, geCDP95, le3.0,
ge4.0, ge8.0

but I get the following error:

ERROR  : yyerror() -> syntax error in file
"/lfs/h2/emc/stmp/deanna.spindler/evs_output/jevs_nfcens_g2o_stats.20023648.cbqs01/tmp_00_fhr1/met_config_186622_0"
ERROR  :
ERROR  :    line   = 52
ERROR  :
ERROR  :    column = 32
ERROR  :
ERROR  :    text   = "geSFP95"
ERROR  :
ERROR  :
ERROR  :   field = [{ name="HTSGW"; level="L0"; cat_thresh=[
le3.0,ge4.0,ge8.0,geSFP95,geSCP95,geCDP95,ge==FBIAS1,==FBIAS1,==FBIAS1 ];
}];
ERROR  :
_______________________________^^^^^^^___________________________________________________________________________________________

Any idea what I did wrong?

Thanks,
Deanna

**Comment by DeannaSpindler-NOAA** [comment]
Hello John,

I got the thresholds to be accepted:
FCST_VAR1_NAME = HTSGW
FCST_VAR1_LEVELS = L0
FCST_VAR1_THRESH = <=3.0, >=4.0, >=8.0, >=SFP95, >=SCP95, >=CDP95, ==FBIAS1, ==FBIAS1, ==FBIAS1

OBS_VAR1_NAME = HTSGW
OBS_VAR1_LEVELS = L0
OBS_VAR1_THRESH = <=3.0, >=4.0, >=8.0, >=SOP95, >=SCP95, >=CDP95, <=3.0, >=4.0, >=8.0

But now I have another puzzle: it works for the 00 cycle with no errors,
but for the 12 cycle it will not use the climatology:
Skipping the 19701101_000000 "HTSGW(0,*,*)" climatology field since the
time offset (43200 seconds) >= the "hour_interval" entry (10800 seconds)
from wave_climo_mean_1128.nc -- which causes there to be errors since it
cannot compute the climate percentiles.
It's the same PointStat config file for both cycles and the same
climatology file.  The climatology has data every 3H, so it has data for
the 12H.
I have POINT_STAT_CLIMO_MEAN_HOUR_INTERVAL = 3 in the config file.

Can you help me, please?

Thanks,
Deanna

**Reply by JohnHalleyGotway** [reply]
Deanna, here's my *guess* based on the log message you sent:
```
Skipping the 19701101_000000 "HTSGW(0,*,*)" climatology field since the
time offset (43200 seconds) >= the "hour_interval" entry (10800 seconds)
from wave_climo_mean_1128.nc
```
You're reading climo data from a NetCDF file, namely wave_climo_mean_1128.nc. I'm guessing that the "HTSGW" NetCDF variable has 3 dimensions, one for time plus 2 spatial dimensions. Based on this `HTSGW(0,*,*)`, it looks like you've configured it to always use the first time from that file (i.e. time index = 0).

When you verify at 00Z, all is well because the data for the first time index matches. When you verify at 12Z, MET reads the 00Z climo data from time index 0 but doesn't use it because it differs from verification time by 12 hours (i.e. 43200 seconds from the log message). I can think of 3 fixes, in order from worst to best.

First, you could try rerunning at 12Z using `HTSGW(3,*,*)` since presumably time index 3 corresponds to 12Z. But that's tedious because you'd need to change that integer time index based on the valid hour for every run!

Second, if you're running met-11.0.0-beta5, try switching from `HTSGW(0,*,*)` to `HTSGW({valid?fmt=%Y%m%d_%H%M%S},*,*)`. So instead of specifying the time index, you define a time string for the METplus wrapper to define based on the current valid time being evaluated. And the MET library code picks out the time index corresponding to that string. This logic was adding for one of the met-11.0.0 beta releases.

Third, we should probably enhance MET to do a better job reading climo data from NetCDF files. With GRIB input files, MET steps through all the GRIB records looking for data before/after the current valid time being evaluated. So if you have climo data every 3 hours, but are evaluating at 07Z, it'll find the 06Z and 09Z data and interpolate between them to 07Z. That functionality does **NOT** work for NetCDF data. The MET library code is not smart enough to step through time slices to find the data before/after the valid time being evaluated.

So I'd recommend that you try the second option, defining the time dimension using the valid time string. But please also think about the third option. Do you see a need for us to enhance MET to handle NetCDF climo data in that way? If so, we should get that written up a new issue in GitHub.

**Comment by DeannaSpindler-NOAA** [comment]
Hello John,

I think you are right, it's only reading my climatology at the 00 cycles
because of how I am specifying it.
Unfortunately this is for EVSv1, so I have to use MET/10.1.2 at most.  I
can pass the climate index to the config file, but I don't know how to
specify it in the config - will this work?
POINT_STAT_CLIMO_MEAN_FIELD =
   { name="HTSGW"; level="(ENV[climo_idx],*,*)"; },

Thanks,
Deanna

--
Deanna Spindler, PhD
Physical Scientist III
Lynker at NOAA/NWS/NCEP/Environmental Modeling Center
NOAA Center for Weather and Climate Prediction

**Reply by JohnHalleyGotway** [reply]
Great news. I was wrong about when we added support for time strings. Turns out it is supported in version 10.1.2. The new feature for 11.0.0 is generalizing that support from only the time dimension to other dimensions as well, as described in dtcenter/MET#1815.

So please proceed with trying this METplus configuration.
```
POINT_STAT_CLIMO_MEAN_FIELD = { name="HTSGW"; level="({valid?fmt=%Y%m%d_%H%M%S},*,*)"; }
```
Does that produce the desired result?

**Comment by DeannaSpindler-NOAA** [comment]
Hello John,

Unfortunately it doesn't work with a climatology file - mine are referenced
to 1970 so the year doesn't match:
DEBUG 4: Met2dDataFileFactory::new_met_2d_data_file() -> created new
Met2dDataFile object of type "FileType_NcCF".
DEBUG 4: NcCfFile::open() -> parsing units for the time variable "hours
since 1970-11-01 00:00:00"
DEBUG 4: parse_cf_time_string() -> parsed NetCDF CF convention time unit
string "hours since 1970-11-01 00:00:00"
DEBUG 4:         as a reference time of 19701101_000000 and 3600 second(s)
per time step.
DEBUG 4: get_nc_data(NcVar *, double *) add_offset = 0, scale_factor=1,
cell_count=8, is_unsigned_value: 0 for time
DEBUG 4: apply_scale_factor(double) unpacked data: count=0 out of 8.
FillValue(int64)=-9999 data range [0 - 21] raw data: [0 - 21] Positive co$
DEBUG 4: NcCfFile::open() -> could not extract init time from the
"forecast_reference_time" variable.
DEBUG 4: NcCfFile::open() -> could not extract init time from file name.
DEBUG 4: VarInfoFactory::new_var_info() -> created new VarInfo object of
type "FileType_NcCF".
WARNING: MetNcCFDataFile::collect_time_offsets(VarInfo &) -> Not found time
out of 8.

I tried passing the complete string from the script:
export climo_level_str='{ name="HTSGW"; level="(4,*,*)"; }'
and this in the config file:
POINT_STAT_CLIMO_MEAN_FIELD = {ENV[climo_level_str]}
but it still doesn't work.  I can't find the reason why not in the log,
there are no errors, just:
DEBUG 2: For HTSGW/L0 found 1 forecast levels, 0 climatology mean levels,
and 0 climatology standard deviation levels.

Best,
Deanna

**Comment by DeannaSpindler-NOAA** [comment, marked as answer]
I got it! passing the whole string does work, I just had an extra blank space after the = sign.
It's running now, and finding the matching climatology hours in the climo mean and sdev files!

**Reply by JohnHalleyGotway** [reply]
That's great. I'm glad you were able to figure it out!

That detail about the timestamp not matching also occurred to me on my drive to work this morning. You certainly can just hard-code the 4 digit year as 1970 rather than using the year of the current valid time:
```
POINT_STAT_CLIMO_MEAN_FIELD = { name="HTSGW"; level="(1970{valid?fmt=%m%d_%H%M%S},*,*)"; }
```
So this would always use 1970 and then append the month, day, year, hour, minute, and second of the valid time being evaluated. I don't know what your NetCDF data actually contains, but if the month and day also remain constant as 1, you'd use:
```
POINT_STAT_CLIMO_MEAN_FIELD = { name="HTSGW"; level="(19700101_{valid?fmt=%H%M%S},*,*)"; }
```
So you'd adjust your configuration based on the time stamps that are actually present in those climo files.

If you're satisfied and don't have further questions, please click the "Mark As Answer" button for one of the previous responses. And then I'll lock this discussion to prevent future posts.
